pkg/controllerutils/patch.go

```diff
@@ -148,3 +150,43 @@ func createOrGetAndPatch(ctx context.Context, c client.Client, obj client.Object
 
 	return controllerutil.OperationResultCreated, nil
 }
+
+// TryPatch tries to apply the given transformation function onto the given object, and to patch it afterwards with optimistic locking.
+// It retries the patch with an exponential backoff.
+func TryPatch(ctx context.Context, backoff wait.Backoff, c client.Client, obj client.Object, transform func() error) error {
+	return tryPatch(ctx, backoff, c, obj, c.Patch, transform)
+}
+
+// TryPatchStatus tries to apply the given transformation function onto the given object, and to patch its
+// status afterwards with optimistic locking. It retries the status patch with an exponential backoff.
+func TryPatchStatus(ctx context.Context, backoff wait.Backoff, c client.Client, obj client.Object, transform func() error) error {
+	return tryPatch(ctx, backoff, c, obj, c.Status().Patch, transform)
+}
```

Comment:
Arrrgggg, this commit is pure evil 😈 
I spent so much time on getting rid of those functions in the context of https://github.com/gardener/gardener/issues/2822 and now you're adding them back??? 😢 
Please see https://github.com/gardener/gardener/pull/4378 / https://github.com/gardener/gardener/blob/master/docs/development/kubernetes-clients.md#dont-retry-on-conflict

The `Try*` functions were left only in the extensions library for legacy reasons and it was planned to remove them later on, see https://github.com/gardener/gardener/issues/2414:

>  Eliminate wrongful usages of Try{Update,Patch}, RetryOnConflict, exponentialBackoff in extensions library

If you don't want to find proper alternatives for `Try*` in the resource manager code right now, fine. Then please add a similar item to https://github.com/gardener/gardener/issues/2822. But please don't add those funcs to `pkg/controllerutils` again – the usage will just spread again and we will be in a similar mess like before my work. 🙏

Reply:
> I spent so much time on getting rid of those functions in the context of #2822 and now you're adding them back??? 😢
> Please see #4378 / https://github.com/gardener/gardener/blob/master/docs/development/kubernetes-clients.md#dont-retry-on-conflict

#2822 did not tackle the g/grm code, and as the goal of this PR is moving g/grm into g/g: Yes, I obviously have to add them again. Also, such usages is still spread in the extensions library, so the functions were basically never gone, just "invisible"/not part of this package. 

> If you don't want to find proper alternatives for Try* in the resource manager code right now, fine. Then please add a similar item to #2822.

I have added another task to https://github.com/gardener/gardener/issues/2414, similar to "Eliminate wrongful usages of `Try{Update,Patch}`, `RetryOnConflict`, `exponentialBackoff` in extensions library".

> But please don't add those funcs to pkg/controllerutils again – the usage will just spread again and we will be in a similar mess like before my work. 🙏

Since both extensions and GRM are using these functions, I have worked on d2e00d945e4e198e2fff0c7156453b75557aa8fd which is removing some duplicate code.
I have added deprecation information to the doc strings to prevent developers from using them, but generally, I doubt that moving them to another package can prevent wrongful usages.